### PR TITLE
fix #379: npm install runs under windows 10

### LINKED ISSUE #379: [Bug] "npm install" fails on windows, msvc 2017, ...
Hi all,

I am coming from trying out the temporial samples for typescript and have filed a [bug](https://github.com/temporalio/samples-typescript/issues/61#issue-1050816961) over there.

After installing all dependencies as described in the readme of this project, running of `npm install` fails on windows 10, msvc 2017, node 16, rust 1.56.

I could identify three issues:
1) packages/core-bridge package.json build-rust script is not cross OS
2) core-bridge scripts/build.js calls a cargo-cp-artifact via OS spawn, which is not cross OS
3) packages/worker compile throws:
```
npm ERR! Die Projekte in dieser Projektmappe werden nacheinander erstellt. Um eine parallele Erstellung zu ermöglichen, müssen Sie den Schalter "-m" hinzufügen.
npm ERR!   workflow-isolate-extension.cc
npm ERR! ~\my-project-name\node_modules\isolated-vm\src\lib\thread_pool.h(17,55): warning C4003: Nicht genügend Argumente für den Aufruf des funktionsähnlichen Makros "max". [~\my-project-name\node_modules\@temporalio\worker\build\temporalio-workflow-isolate-extension.vcxproj]       
npm ERR! ~\my-project-name\node_modules\isolated-vm\src\lib\thread_pool.h(17,55): error C2589: "(": Ungültiges Token auf der rechten Seite von "::" [~\my-project-name\node_modules\@temporalio\worker\build\temporalio-workflow-isolate-extension.vcxproj]
npm ERR! ~\my-project-name\node_modules\isolated-vm\src\lib\thread_pool.h(17): error C2062: "unknown-type"-Typ unerwartet [~\my-project-name\node_modules\@temporalio\worker\build\temporalio-workflow-isolate-extension.vcxproj]
npm ERR! ~\my-project-name\node_modules\isolated-vm\src\lib\thread_pool.h(17,22): error C2059: Syntaxfehler: ")" [~\my-project-name\node_modules\@temporalio\worker\build\temporalio-workflow-isolate-extension.vcxproj]
```

### To Reproduce

setup Windows 10 (21H1), 
node 16 (via nvm install 16 && nvm use 16), 
rustup (cargo 1.56, rustc 1.56.1), 
vs 2017 (via choco visualstudio2017-workload-vctools visualstudio2017buildtools)

### Expected behavior

`npm install` and `npm test` run on Windows environment successfully (with the appropriate dependencies)

### PR DESCRIPTION
**Pull request by cons0l3**
closes #379 
tested on win10 and under wsl (ubuntu 20)

### CODE REVIEW

File: packages/core-bridge/package.json [hunk 1]
```diff
@@ -5,7 +5,7 @@
   "main": "index.node",
   "types": "index.d.ts",
   "scripts": {
-    "build-rust": "TEMPORAL_WORKER_FORCE_BUILD=true node ./scripts/build.js",
+    "build-rust": "npx cross-env TEMPORAL_WORKER_FORCE_BUILD=true node ./scripts/build.js",
```
Comment (bergundy) [inline packages/core-bridge/package.json:8]: I don't think you need `npx` in package scripts, the path should already be set here.

File: packages/core-bridge/scripts/build.js [hunk 2]
```diff
@@ -39,28 +40,13 @@ function compile(target) {
...
-      '--release',
-      ...(target ? ['--target', target] : []),
-    ],
-    { stdio: 'inherit' }
-  );
-  if (error !== undefined) {
+
+  try {
+    argv = ['--artifact', 'cdylib', 'temporal_sdk_typescript_bridge', out, '--', 'cargo', 'build', '--message-format=json-render-diagnostics', '--release', ...(target ? ['--target', target] : []), ]
+    cargo_cp_artifact(argv, process.env)
```
Comment (bergundy) [inline packages/core-bridge/scripts/build.js:46]: After reading the code for this package: https://github.com/neon-bindings/cargo-cp-artifact/blob/main/src/index.js
I'm convinced you should await here.

File: packages/worker/binding.gyp [hunk 3]
```diff
@@ -9,14 +9,22 @@
...
-        'VCCLCompilerTool': { 'ExceptionHandling': 1 },
+        'VCCLCompilerTool': { 'ExceptionHandling': 1, 'AdditionalOptions': [ '/GR' ] },
       },
+      'msvs_disabled_warnings': [
+					4101, # Unreferenced local (msvc fires these for ignored exception)
+					4068, # Unknown pragma
+				],
+			'conditions': [
+					[ 'OS == "win"', { 'defines': [ 'NOMSG', 'NOMINMAX', 'WIN32_LEAN_AND_MEAN' ] } ],
+				],
```
Comment (bergundy) [inline packages/worker/binding.gyp:27]:
```suggestion
      'msvs_disabled_warnings': [
        4101, # Unreferenced local (msvc fires these for ignored exception)
        4068, # Unknown pragma
      ],
      'conditions': [
        [ 'OS == "win"', { 'defines': [ 'NOMSG', 'NOMINMAX', 'WIN32_LEAN_AND_MEAN' ] } ],
      ],      
```

File: packages/core-bridge/package.json (hunk 1 above)
Comment (bergundy) [reply packages/core-bridge/package.json:8]: At this point I'd avoid using env vars and switch to command line arguments.

**Review by bergundy — COMMENTED** [review]
Thanks for this.
I'll want to add a Windows builder in GH actions to test all of this.
If you want to contribute that too, great but otherwise I can take it from here.
I can try to push it in this week.

File: packages/core-bridge/package.json (hunk 1 above)
Comment (cons0l3) [reply packages/core-bridge/package.json:8]: good idea to switch to command line args. I have seen several env variables in build.js starting with "TEMPORAL_". they should all be switched. I can work on that tomorrow.

File: packages/core-bridge/scripts/build.js (hunk 2 above)
Comment (cons0l3) [reply packages/core-bridge/scripts/build.js:46]: I am not a js pro, but I disagree. The exported run() is not declared async. The "bin/.." script does not use an await itself... But I could be wrong ;-)

Comment (bergundy) [reply packages/core-bridge/scripts/build.js:46]: Right, I took another look and it seems like this tool isn't meant to be imported.
You have to run it as a child process.

```js
spawnSync(process.argv[0], [which.sync('cargo-cp-artifact'), /* rest of the args */]`)
```
https://www.npmjs.com/package/which

File: packages/core-bridge/package.json (hunk 1 above)
Comment (cons0l3) [reply packages/core-bridge/package.json]: removed environment variables TEMPORAL_WORKER_FORCE_BUILD and TEMPORAL_WORKER_BUILD_TARGETS with argparse command line options. there is no fallback to the env variables.

I do not know which impact this has for the CI/CD pipeline(s). @bergundy: you have to take over.

```
.> node scripts/build.js -h
usage: build.js [-h] [-v] [-t {x86_64-apple-darwin,aarch64-apple-darwin,x86_64-unknown-linux-gnu,aarch64-unknown-linux-gnu,x86_64-pc-windows-gnu,all}] [-f]

build.js compiles the sdk-core via cargo and then packages it as javascript module

optional arguments:
  -h, --help            show this help message and exit
  -v, --version         show program's version number and exit
  -t {x86_64-apple-darwin,aarch64-apple-darwin,x86_64-unknown-linux-gnu,aarch64-unknown-linux-gnu,x86_64-pc-windows-gnu,all}, --targets {x86_64-apple-darwin,aarch64-apple-darwin,x86_64-unknown-linux-gnu,aarch64-unknown-linux-gnu,x86_64-pc-windows-gnu,all}
                        specifies the targets to build for (formerly env.TEMPORAL_WORKER_BUILD_TARGETS)
  -f, --force           forces a clean rebuild of sdk-core and repackaging (formerly env.TEMPORAL_WORKER_FORCE_BUILD)

Happy compiling!
```

File: packages/core-bridge/scripts/build.js (hunk 2 above)
Comment (cons0l3) [reply packages/core-bridge/scripts/build.js]: used "which" and "spawnSync", removed direct call

**Comment by bergundy** [comment]
Thanks a lot @cons0l3 I'll take it from here and make sure it's tested

**Pull request closed without merging**